src/worker/jobs/delete_crate.rs

```diff
@@ -0,0 +1,50 @@
+use crate::storage::FeedId;
+use crate::worker::Environment;
+use anyhow::Context;
+use crates_io_worker::BackgroundJob;
+use std::sync::Arc;
+use tokio::try_join;
+
+/// A background job that deletes all files associated with a crate from the storage backend.
+#[derive(Serialize, Deserialize)]
+pub struct DeleteCrateFromStorage {
+    name: String,
+}
+
+impl DeleteCrateFromStorage {
+    pub fn new(name: String) -> Self {
+        Self { name }
+    }
+}
+
+impl BackgroundJob for DeleteCrateFromStorage {
+    const JOB_NAME: &'static str = "delete_crate_from_storage";
+
+    type Context = Arc<Environment>;
+
+    async fn run(&self, ctx: Self::Context) -> anyhow::Result<()> {
+        let name = &self.name;
+
+        try_join!(
+            async {
+                info!("{name}: Deleting crate files from S3…");
+                let result = ctx.storage.delete_all_crate_files(name).await;
+                result.context("Failed to delete crate files from S3")
+            },
+            async {
+                info!("{name}: Deleting readme files from S3…");
+                let result = ctx.storage.delete_all_readmes(name).await;
+                result.context("Failed to delete readme files from S3")
+            },
+            async {
+                info!("{name}: Deleting RSS feed from S3…");
+                let feed_id = FeedId::Crate { name };
+                let result = ctx.storage.delete_feed(&feed_id).await;
+                result.context("Failed to delete RSS feed from S3")
+            }
+        )?;
```

Comment:
is it sufficient for this to be in one job or should it be split into multiple jobs that can be retried individually if they fail? 🤔

Reply:
Yeah, what would happen if one of these failed?

Reply:
`try_join!()` cancels the other runnings tasks if one of them fails. it then returns the error, which will make the job fail and will cause it to be retried a little later. AFAIU all of the steps in this job are idempotent though, so retrying isn't a big deal. the job does not perform CDN invalidations yet though.

Reply:
Let's say task 1 succeeds first, but task 2 fails, so it cancels task 3. What will the state of task 1 be after all? If it remains done, is it safe to retry?

Reply:
I think if every job could be retried even if it's already completed, this wouldn't be a problem.

Reply:
> If it remains done, is it safe to retry?

yes, as far as I understand this shouldn't be an issue. I guess splitting it up is probably unnecessary over-engineering 😅